experimental: Integrate PropertyLayers into cell space

EwoutH opened the pull request:
Integrate the PropertyLayers into the cell space. Initially only in the DiscreteSpace, and thereby the Grids.

Since we're in the experimental space, this PR is ready for review as-is.

In future PRs:
- Add a more comprehensive select function
https://github.com/projectmesa/mesa/blob/4c829896975f57ceffaf6b403e89fac25490b04d/mesa/space.py#L872
- Integrate PropertyLayer into CellCollection
  - Thinking about how to do this fast. For larger collections, keeping a mask is probably most performant, but for smaller collections, maybe just iterate though the cells.

quaquel commented on mesa/experimental/cell_space/cell.py:
```diff
@@ -191,3 +194,30 @@ def _neighborhood(
             return neighborhood
+
+    # PropertyLayer methods
+    def add_property_layer(self, property_layer: PropertyLayer):
```
do add and remove an entire layer belong in a cell?

In my view a cell would just do a get and a set on its coordinates in a specified layer.

EwoutH replied:
Having direct access to the PropertyLayer is what makes it performant. You don't want to copy all the values everything. So normally you directly access the layer at your coordinate.

EwoutH replied:
Removed those two functions, you can directly access the dicts.

quaquel commented on mesa/experimental/cell_space/discrete_space.py:
```diff
@@ -73,3 +83,38 @@ def empties(self) -> CellCollection[T]:
+        self, property_name: str, value, condition: Callable[[T], bool] | None = None
+    ):
+        """Set the value of a property for all cells in the grid."""
+        self.property_layers[property_name].set_cells(value, condition)
```
please (for all) specify the arguments as well as passing a docstring.

in pycharm, you can specify which docstring standard you want to use to make this easy (https://www.jetbrains.com/help/pycharm/settings-tools-python-integrated-tools.html)

EwoutH replied:
First of all, very useful tool. And you know I love good tooling, so thanks for linking to it.

> (for all) specify the arguments

I don't know if I necessarily agree with this. Every function, method and class should have at least the oneliner and type-hinted input and output. But I also think that sometimes can be enough.

Comments/docstring are there to:
- Explain conceptual or higher level workings
- Explain unintuitive or complex code.

If there are no conceptual workings and the code is simple and intuitive, it's readable on its own. So I don't think it needs docstring in those cases.

I will reflect on these specific cases.

EwoutH replied:
Added quite some docstring

quaquel commented:
It's remarkably straightforward. Good to see.

EwoutH commented:
If there are any other remarks please let me know.

Corvince commented:
> It's remarkably straightforward. Good to see.

I can second that. When I first skimmed this PR I thought its not ready yet, because its not a lot of code. Would have expected more work to be needed. Great to see! One thing I need more time to think about is how the individual cell interacts with the layer. Right now a reference to the whole Property layer ist passed. But Cells also have a "properties" attribute. I would have imagined that inside this properties dictionary there is just a "view" to the propertylayers individual coordinate. Does that make sense?

quaquel commented:
One last small request: 4 lines are currently not covered by the tests. Can you make sure we have 100% coverage on this? Its a small thing to fix now, but it adds up towards the future.

EwoutH commented:
Agreed, feel free to add them, otherwise I will pick it up later this week.

EwoutH commented:
Updated the tests, and removed the possibility to initialize grids with propertylayers, because that doesn't give you the control if you want those layers added to the cells or not.

I'm struggeling a bit with how - on a conceptual level - we want Cell `poperties` and `property_layers` to interact / coexist / complement each other. Functionally it's clear, `poperties` are for individual cell properties (also other than numbers) and `property_layers` are for high-performance properties that are shared though a whole space.

Input is welcome.

quaquel commented:
> I'm struggeling a bit with how - on a conceptual level - we want Cell poperties and property_layers to interact / coexist / complement each other. Functionally it's clear, poperties are for individual cell properties (also other than numbers) and property_layers are for high-performance properties that are shared though a whole space.

I don't understand the point, but also in the code, I am a bit lost. In my understanding, property layers are define at the space level and individual cells can only access their values in those same property layers. However it seems cell now has its own property layer dict?

quaquel commented on mesa/experimental/cell_space/cell.py:
```diff
@@ -69,6 +71,7 @@ def __init__(
         self.capacity: int = capacity
         self.properties: dict[Coordinate, object] = {}
         self.random = random
+        self.property_layers: dict[str, PropertyLayer] = {}
```
sorry I don't understanand why this is here? should this not be linked to the property layers defined at the grid level?

EwoutH replied:
It also is, but it might be useful to be able to access your value in a PropertyLayer directly from the cell.

quaquel replied:
I understand that and agree with it. I now also looked at `add_property_layer` and start to understand what is going on. 

1. I suggest making this `self._property_layers` so it's not declared as part of the public API. My concern is that one might add a layer via cell directly which would then not exist on the grid. 
2. I think overall, a more elegant solution is possible here (but that can be a separate PR). My idea would be to add on the fly Descriptors to cells for each property layer. This would make it possible to de `cell.property_layer_name`, which would return the value for this cell in the named property layer. Another benefit would be that we don't have code by default in cell that is only functional in OrthogonalGrids.

Corvince replied:
On 2) my high-level idea would be that we use the existing property dictionary of the cell and add to it an entry with the name of the property layer, so `cell.properties[property_layer_name]`. *getting* the value should be easy, but I don't know how that would work for setting the value. Any ideas?

quaquel replied:
I can't think of an immediate solution other than to subclass dict and have a dedicated PropertiesDict for all this.

EwoutH replied:
Was thinking about that. It might be great as an API, but can also create naming conflicts.

Unfortunately, I'm really making a sprint on my thesis now, so not much conceptual-thought capacity to work on this.

quaquel replied:
With the one minor change (to a private attribute), I think we can merge this for now. I might have some time over the weekend to play around with more descriptor fun 😄 and see what I can do.

EwoutH replied:
Go ahead in making the change and merging!
